### PR TITLE
feat: Add ruby code - chapter "Hashing"

### PR DESCRIPTION
**Pull request by khoaxuantu**
If this pull request (PR) is associated with **coding or code transpilation**, please attach the relevant console outputs to the PR and complete the following checklist:

- [x] I have thoroughly reviewed the code, focusing on its formatting, comments, indentation, and file headers.
- [x] I have confirmed that the code execution outputs are consistent with those produced by the reference code (Python or Java).
- [x] The code is designed to be compatible on standard operating systems, including Windows, macOS, and Ubuntu.

---

chapter_hashing_1
chapter_hashing_2
chapter_hashing_3
chapter_hashing_4

### CODE REVIEW

File: codes/ruby/chapter_hashing/array_hash_map.rb [hunk 1]
```diff
@@ -0,0 +1,120 @@
+=begin
+File: array_hash_map.rb
+Created Time: 2024-04-13
+Author: Xuan Khoa Tu Nguyen (ngxktuzkai2000@gmail.com)
+=end
+
+### 建值对 ###
```
Comment (bluebean-cloud) [inline codes/ruby/chapter_hashing/array_hash_map.rb:7]: 建值对 => 键值对

File: codes/ruby/chapter_hashing/array_hash_map.rb [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+class Pair
+  attr_accessor :key, :val
+
+  def initialize(key, val)
+    @key = key
+    @val = val
+  end
+end
+
+### 基丁数组实现的哈希表 ###
```
Comment (bluebean-cloud) [inline codes/ruby/chapter_hashing/array_hash_map.rb:17]: 基丁 => 基于

File: codes/ruby/chapter_hashing/array_hash_map.rb [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+  hmap.put(16750, "小算")
+  hmap.put(13276, "小法")
+  hmap.put(10583, "小鸭")
+  puts "\n添加完成后，哈希表为\nKey -> Value"
+  hmap.print
+
+  # 查询操作
+  # 向哈希表中输入键 key , 得到值 value
+  name = hmap.get(15937)
+  puts "\n输入学好 15937 ，查询到姓名 #{name}"
```
Comment (bluebean-cloud) [inline codes/ruby/chapter_hashing/array_hash_map.rb:97]: 输入学好 => 输入学号

File: codes/ruby/chapter_hashing/hash_map_open_addressing.rb [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+    nil
+  end
+
+  ### 添加操作 ###
+  def put(key, val)
+    # 当负载因子超过阈值时，执行扩容
+    extend if load_factor > @load_thres
+    # 搜索 key 对应的桶索引
+    index = find_bucket(key)
+    # 若找到键值对，则覆盖开返回
```
Comment (bluebean-cloud) [inline codes/ruby/chapter_hashing/hash_map_open_addressing.rb:73]: 则覆盖开返回 => 则覆盖 val 并返回

File: codes/ruby/chapter_hashing/hash_map_open_addressing.rb [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+    # 搜索 key 对应的桶索引
+    index = find_bucket(key)
+    # 若找到键值对，则用删除标记覆盖它
+    unless [nil, TOMBSTONE].include?(@buckets[index])
+      @buckets[index] = TOMBSTONE
+      @size -= 1
+    end
+  end
+
+  ### 扩容海西表 ###
```
Comment (bluebean-cloud) [inline codes/ruby/chapter_hashing/hash_map_open_addressing.rb:94]: 扩容海西表 => 扩容哈希表

File: codes/ruby/chapter_hashing/array_hash_map.rb (hunk 1 above)
Comment (bluebean-cloud) [inline codes/ruby/chapter_hashing/array_hash_map.rb:7]: There is a typo here.

```suggestion
### 键值对 ###
```

File: codes/ruby/chapter_hashing/array_hash_map.rb (hunk 2 above)
Comment (bluebean-cloud) [inline codes/ruby/chapter_hashing/array_hash_map.rb:17]: There is a typo here.

```suggestion
### 基于数组实现的哈希表 ###
```

File: codes/ruby/chapter_hashing/array_hash_map.rb (hunk 3 above)
Comment (bluebean-cloud) [inline codes/ruby/chapter_hashing/array_hash_map.rb:97]: There is a typo here.

```suggestion
  puts "\n输入学号 15937 ，查询到姓名 #{name}"
```

Comment (khoaxuantu) [reply codes/ruby/chapter_hashing/array_hash_map.rb:97]: > There is a typo here.

It will be resolved. You are repeating the reviews. It is unnecessary

File: codes/ruby/chapter_hashing/hash_map_chaining.rb [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+    index = hash_func(key)
+    bucket = @buckets[index]
+    # 遍历桶，若遇到指定 key ，则更新对应 val 并返回
+    for pair in bucket
+      if pair.key == key
+        pair.val = val
+        return
+      end
+    end
+    # 若无该 key ，则将建值对添加至尾部
```
Comment (bluebean-cloud) [inline codes/ruby/chapter_hashing/hash_map_chaining.rb:55]: There is a typo here.

```suggestion
    # 若无该 key ，则将键值对添加至尾部
```

File: codes/ruby/chapter_hashing/hash_map_chaining.rb [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+
+  ### 扩容哈希表 ###
+  def extend
+    # 暫存原哈希表
+    buckets = @buckets
+    # 初始化扩容后的新哈希表
+    @capacity *= @extend_ratio
+    @buckets = Array.new(@capacity) { [] }
+    @size = 0
+    # 将建值对从原哈希表搬运至新哈希表
```
Comment (bluebean-cloud) [inline codes/ruby/chapter_hashing/hash_map_chaining.rb:83]: There is a typo here.

```suggestion
    # 将键值对从原哈希表搬运至新哈希表
```

File: codes/ruby/chapter_hashing/hash_map_open_addressing.rb [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+
+  ### 搜索 key 对应的桶索引 ###
+  def find_bucket(key)
+    index = hash_func(key)
+    first_tombstone = -1
+    # 线性探测，当遇到空桶时跳出
+    while !@buckets[index].nil?
+      # 若遇到 key ，返回对应的桶索引
+      if @buckets[index].key == key
+        # 若之前遇到了删除标记，则将建值对移动至该索引处
```
Comment (bluebean-cloud) [inline codes/ruby/chapter_hashing/hash_map_open_addressing.rb:40]: There is a typo here.

```suggestion
        # 若之前遇到了删除标记，则将键值对移动至该索引处
```

**Comment by khoaxuantu** [comment]
@bluebean-cloud I am aware of those typos and am resolving it. But I see here you're repeating yourself the reviews. It is unnecessary and make the PR hard to follow. Please wrap all the reviews in one comment only.

**Review by bluebean-cloud — CHANGES_REQUESTED** [review]
Hi @khoaxuantu , thanks for the PR, I have some comments

File: codes/ruby/chapter_hashing/array_hash_map.rb (hunk 3 above)
Comment (bluebean-cloud) [reply codes/ruby/chapter_hashing/array_hash_map.rb:97]: I'm sorry for that 😭 

I'm not familiar with github's workflow, I apologize for my mistake

**Comment by krahets** [comment]
Hi @khoaxuantu , did you manually type the code comments? I recommend you copy&paste all the comments to keep them consistent. Or you may double check the code with GPT.

**Comment by khoaxuantu** [comment]
> Hi @khoaxuantu , did you manually type the code comments? I recommend you copy&paste all the comments to keep them consistent. Or you may double check the code with GPT.

I tried to manually type the comments in this PR. Would have been checked carefully later on 👌
